### DISCUSSION TITLE
Issues with Viewport SetContent

### DISCUSSION THREAD

**Discussion by jamesthesken** [charmbracelet/bubbletea#374] [Q&A] [answered]
Hey Charm team, thank you for this library as it's been very inspiring and fun to work with!

I am working on an IRC client, and having some trouble updating the Viewport component: 
https://github.com/jamesthesken/go-irc/tree/tui_v1

**The Issue**

In `go-irc/tui/tui.go` there is a function called `StartTea()`, which starts a goroutine that listens for incoming messages from an IRC server.

This function, [`Read()`](https://github.com/jamesthesken/go-irc/blob/f815fd7c256eba7982ba0dba500a61442675a183/tui/tui.go#L150) calls `model.Update(RcvMessage{content: msgRcv})`, where `RcvMessage` is defined as:

```
type RcvMessage struct {
	content string
}
```

After this message is received it is processed by a switch statement in [`Update()`](https://github.com/jamesthesken/go-irc/blob/f815fd7c256eba7982ba0dba500a61442675a183/tui/tui.go#L97):

```
case RcvMessage:
		// this does not work:
		m.messages = append(m.messages, m.senderStyle.Render("Server: ")+string(msg.content))
		m.viewport.SetContent(strings.Join(m.messages, "\n"))
		fmt.Print(msg.content) // but this does ??
		m.viewport.GotoBottom()
```

I can't seem to figure it out, but I can print the message contents to the screen, yet setting content in the Viewport does not work as expected. Even with a dummy string (I.e., `m.viewport.SetContent("received")`

Could there be a race condition between my message retrieval goroutine and updating the viewport? 

Thank you in advance for any help and insight!

**Comment by muesli** [comment, marked as answer]
You shouldn't call `model.Update` directly: either send a message to the `tea.Program` or wrap the IRC handling in a `tea.Command`. Bubble Tea will then handle your messages, call `Update` and re-render the state of your program.

**Comment by jamesthesken** [comment]
> You shouldn't call `model.Update` directly: either send a message to the `tea.Program` or wrap the IRC handling in a `tea.Command`. Bubble Tea will then handle your messages, call `Update` and re-render the state of your program.

Thank you so much! That is a great feature. I got it working by sending a message to the tea program:

```
func (m Model) Read(conn io.ReadWriter, p *tea.Program) {
	s := bufio.NewScanner(conn)
	for s.Scan() {
		line := s.Text()
		msgRcv := fmt.Sprintf(client.ParseMessage(line), "\n")
		p.Send(RcvMessage{content: msgRcv})
	}
	if s.Err() != nil {
		log.Fatalf("Error occured: %s", s.Err())
	}
}
```

**Reply by jamesthesken** [reply]
In case anyone comes back to this discussion, I think another useful example would be: https://github.com/charmbracelet/bubbletea/blob/master/examples/realtime/main.go

Rather than spinning up a goroutine to listen for incoming messages
